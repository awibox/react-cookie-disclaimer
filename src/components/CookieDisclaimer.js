import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { LocalStorage } from 'combo-storage';

const cookieDisclaimerStyle = {
  position: 'fixed',
  right: 0,
  left: 0,
  boxShadow: '0 0 3px rgba(0, 0, 0, 0.12), 0 0 2px rgba(0, 0, 0, 0.24)',
};

const cookieDisclaimerCloseStyle = {
  textAlign: 'center',
  position: 'absolute',
  transform: 'rotate(45deg)',
  cursor: 'pointer',
  fontFamily: '-webkit-pictograph',
};

class CookieDisclaimer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      close: false,
      hasStorage: LocalStorage.get('react-cookie-disclaimer'),
    };
    this.closeDisclaimer = this.closeDisclaimer.bind(this);
  }

  static propTypes = {
    background: PropTypes.string,
    bottomPosition: PropTypes.bool,
    color: PropTypes.string,
    cookiePolicyLink: PropTypes.string,
    cookiePolicyName: PropTypes.string,
    cookiePolicyText: PropTypes.string,
    closeIconSize: PropTypes.number,
    closeIconPositionTop: PropTypes.bool,
    padding: PropTypes.number,
    text: PropTypes.string,
  };

  static defaultProps = {
    background: '#fff',
    bottomPosition: false,
    color: '#000',
    cookiePolicyLink: '',
    cookiePolicyName: 'Cookie Policy',
    cookiePolicyText: 'By continuing to use the service, you agree to our use of cookies as described in the',
    closeIconSize: 28,
    closeIconPositionTop: false,
    padding: 20,
    text: 'This website uses cookies to improve service, for analytical and advertising purposes.',
  };

  closeDisclaimer() {
    this.setState({ close: true });
    LocalStorage.set('react-cookie-disclaimer', true);
    if (!this.props.bottomPosition) {
      document.body.style.marginTop = 0;
    }
  }

  componentDidMount() {
    if (!this.state.hasStorage && !this.props.bottomPosition) {
      const disclaimer = document.getElementById('cookie-disclaimer').offsetHeight;
      document.body.style.marginTop = `${disclaimer}px`;
    }
  }

  render() {
    if (this.state.hasStorage) {
      return null;
    }
    const {
      background,
      bottomPosition,
      color,
      cookiePolicyLink,
      cookiePolicyText,
      cookiePolicyName,
      closeIconSize,
      closeIconPositionTop,
      padding,
      text,
    } = this.props;
    const closeDisclaimer = this.state.close;
    const positionObj = bottomPosition ? { bottom: 0 } : { top: 0 };
    const styleForCookieDisclaimer = {
      visibility: closeDisclaimer ? 'hidden' : 'visible',
      backgroundColor: background,
      padding: `${padding}px`,
      paddingRight: `${closeIconSize + padding}px`,
      color,
      ...positionObj,
      ...cookieDisclaimerStyle,
    };
    const styleForCloseButton = {
      lineHeight: `${closeIconSize}px`,
      width: `${closeIconSize}px`,
      height: `${closeIconSize}px`,
      fontSize: `${closeIconSize}px`,
      marginTop: closeIconPositionTop ? 0 : `${closeIconSize / -2}px`,
      right: `${padding / 2}px`,
      top: closeIconPositionTop ? `${padding / 2}px` : '50%',
      ...cookieDisclaimerCloseStyle,
    };
    return (
      <div id='cookie-disclaimer' style={styleForCookieDisclaimer}>
        <div id='cookie-disclaimer-close' onClick={this.closeDisclaimer} style={styleForCloseButton}>+</div>
        <div id='cookie-disclaimer-text'>
          {text}
          <span id='cookie-policy' style={{ display: cookiePolicyLink ? 'inline' : 'none' }}>
            {` ${cookiePolicyText} `}
            <a id='cookie-policy-link' href={cookiePolicyLink} style={{ color }}>{cookiePolicyName}</a>
          </span>
        </div>
      </div>
    );
  }
}
export default CookieDisclaimer;
