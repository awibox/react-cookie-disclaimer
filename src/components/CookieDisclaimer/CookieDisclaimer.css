.cookie-disclaimer {
    position: fixed;
    right: 0;
    left: 0;
    padding: 20px 40px 20px 20px;
    box-sizing: border-box;
    -webkit-box-shadow: 0 0 3px rgba(0, 0, 0, 0.12), 0 0 2px rgba(0, 0, 0, 0.24);
    box-shadow: 0 0 3px rgba(0, 0, 0, 0.12), 0 0 2px rgba(0, 0, 0, 0.24);
}
.cookie-disclaimer__close {
    text-align: center;
    line-height: 28px;
    width: 28px;
    height: 28px;
    position: absolute;
    right: 5px;
    top: 50%;
    margin-top: -14px;
    transform: rotate(45deg);
    cursor: pointer;
}
.cookie-disclaimer__close:after {
    display: block;
    font-size: 28px;
    font-family: -webkit-pictograph;
    content: '+';
    font-weight: 100;
}
