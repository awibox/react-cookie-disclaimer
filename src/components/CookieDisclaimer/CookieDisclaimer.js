import React, { Component } from 'react';
import PropTypes from 'prop-types';
// Styles
import './CookieDisclaimer.css';

export class CookieDisclaimer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      close: false,
    };
    this.closeDisclaimer = this.closeDisclaimer.bind(this);
  }

  static propTypes = {
    text: PropTypes.string
  };

  static defaultProps = {
    text: 'We use cookies to operate the website and platform, for analytical purposes, and for advertising/targeting purposes.'
  };

  closeDisclaimer(){
    console.log('Close')
  }

  render() {

    return(
        <div className='cookie-disclaimer'>
          <div className='cookie-disclaimer__close' onClick={this.closeDisclaimer} />
          {this.props.text}
        </div>
    );
  }
}
