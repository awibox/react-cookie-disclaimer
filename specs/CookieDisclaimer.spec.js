import React from 'react';
import { shallow } from 'enzyme';
import CookieDisclaimer from '../src/components/CookieDisclaimer';

const testColor = '#fff';
const testBackground = '#000';
const testText = 'This is correct text';
const testCookiePolicyLink = '/cookie-policy';
const testCookiePolicyName = 'Private Policy';
const testCookiePolicyText = 'This is text before link';
const testIconSize = 22;
const testPadding = 18;
const testCustomContent = 'Custom content';
const testClassName = 'customClass';

describe('CookieDisclaimer', () => {
  it('should render correctly', () => {
    const component = shallow(<CookieDisclaimer/>);
    expect(component).toMatchSnapshot();
  });
  it('should not render', () => {
    const component = shallow(<CookieDisclaimer/>);
    component.setState({hasStorage: true});
    expect(component.get(0)).toEqual(null);
  });
  it('should have background', () => {
    const component = shallow(<CookieDisclaimer background={testBackground}/>);
    expect(component.find('#cookie-disclaimer').get(0).props.style.backgroundColor).toEqual(testBackground);
  });
  it('should have text color', () => {
    const component = shallow(<CookieDisclaimer color={testColor}/>);
    expect(component.find('#cookie-disclaimer').get(0).props.style.color).toEqual(testColor);
  });
  it('should have link color', () => {
    const component = shallow(<CookieDisclaimer color={testColor}/>);
    expect(component.find('#cookie-policy-link').get(0).props.style.color).toEqual(testColor);
  });
  it('should have bottom position', () => {
    const component = shallow(<CookieDisclaimer bottomPosition={true}/>);
    expect(component.find('#cookie-disclaimer').get(0).props.style.bottom).toEqual(0);
  });
  it('should have top position', () => {
    const component = shallow(<CookieDisclaimer/>);
    expect(component.find('#cookie-disclaimer').get(0).props.style.top).toEqual(0);
  });
  it('should hide cookie policy link', () => {
    const component = shallow(<CookieDisclaimer/>);
    expect(component.find('#cookie-policy').get(0).props.style.display).toEqual('none');
  });
  it('should display cookie policy link', () => {
    const component = shallow(<CookieDisclaimer cookiePolicyLink={testCookiePolicyLink}/>);
    expect(component.find('#cookie-policy').get(0).props.style.display).toEqual('inline');
  });
  it('should have correct cookie policy link', () => {
    const component = shallow(<CookieDisclaimer cookiePolicyLink={testCookiePolicyLink}/>);
    expect(component.find('#cookie-policy-link').get(0).props.href).toEqual(testCookiePolicyLink);
  });
  it('should have correct name for cookie policy link', () => {
    const component = shallow(<CookieDisclaimer
      cookiePolicyLink={testCookiePolicyLink}
      cookiePolicyName={testCookiePolicyName}/>);
    expect(component.find('#cookie-policy-link').get(0).props.children).toEqual(testCookiePolicyName);
  });
  it('should have correct text before cookie policy link', () => {
    const component = shallow(<CookieDisclaimer
      cookiePolicyText={testCookiePolicyText}
      cookiePolicyLink={testCookiePolicyLink}
      cookiePolicyName={testCookiePolicyName}/>);
    expect(component.find('#cookie-policy').get(0).props.children[0]).toEqual(` ${testCookiePolicyText} `);
  });
  it('should have correct text for notification', () => {
    const component = shallow(<CookieDisclaimer text={testText}/>);
    expect(component.find('#cookie-disclaimer-text').get(0).props.children[0]).toEqual(testText);
  });
  it('should close notification', () => {
    const component = shallow(<CookieDisclaimer/>);
    component.find('#cookie-disclaimer-close').simulate('click');
    expect(component.find('#cookie-disclaimer').get(0).props.style.visibility).toEqual('hidden');
  });
  it('should close notification bottom position', () => {
    const component = shallow(<CookieDisclaimer bottomPosition={true}/>);
    component.find('#cookie-disclaimer-close').simulate('click');
    expect(component.find('#cookie-disclaimer').get(0).props.style.visibility).toEqual('hidden');
  });
  it('should have correct icon size', () => {
    const component = shallow(<CookieDisclaimer closeIconSize={testIconSize}/>);
    expect(component.find('#cookie-disclaimer-close').get(0).props.style.fontSize).toEqual(`${testIconSize}px`);
    expect(component.find('#cookie-disclaimer-close').get(0).props.style.lineHeight).toEqual(`${testIconSize}px`);
    expect(component.find('#cookie-disclaimer-close').get(0).props.style.width).toEqual(`${testIconSize}px`);
    expect(component.find('#cookie-disclaimer-close').get(0).props.style.height).toEqual(`${testIconSize}px`);
  });
  it('should have center icon position', () => {
    const component = shallow(<CookieDisclaimer
      closeIconSize={testIconSize}
      padding={testPadding}/>);
    expect(component.find('#cookie-disclaimer-close').get(0).props.style.top).toEqual('50%');
    expect(component.find('#cookie-disclaimer-close').get(0).props.style.right).toEqual(`${testPadding / 2}px`);
    expect(component.find('#cookie-disclaimer-close').get(0).props.style.marginTop).toEqual(`${testIconSize / -2}px`);
  });
  it('should have top icon position', () => {
    const component = shallow(<CookieDisclaimer
      closeIconPositionTop={true}
      closeIconSize={testIconSize}
      padding={testPadding}/>);
    expect(component.find('#cookie-disclaimer-close').get(0).props.style.top).toEqual(`${testPadding / 2}px`);
    expect(component.find('#cookie-disclaimer-close').get(0).props.style.marginTop).toEqual(0);
  });
  it('should have correct padding', () => {
    const component = shallow(<CookieDisclaimer
      closeIconSize={testIconSize}
      padding={testPadding}/>);
    expect(component.find('#cookie-disclaimer').get(0).props.style.padding).toEqual(`${testPadding}px`);
    expect(component.find('#cookie-disclaimer').get(0).props.style.paddingRight).toEqual(`${testIconSize + testPadding}px`);
  });
  it('should have custom content', () => {
    const component = shallow(<CookieDisclaimer>{testCustomContent}</CookieDisclaimer>);
    expect(component.find('#cookie-disclaimer').get(0).props.children[1]).toEqual(testCustomContent);
    expect(component.find('#cookie-disclaimer-text').get(0)).toEqual(undefined);
  });
  it('should have a className', () => {
    const component = shallow(<CookieDisclaimer className={testClassName}>
      {testCustomContent}
    </CookieDisclaimer>);
    expect(component.find(`.${testClassName}`)).toExist();
  });
});
