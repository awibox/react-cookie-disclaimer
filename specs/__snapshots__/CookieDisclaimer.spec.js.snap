// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookieDisclaimer should render correctly 1`] = `
<div
  id="cookie-disclaimer"
  style={
    Object {
      "backgroundColor": "#fff",
      "boxShadow": "0 0 3px rgba(0, 0, 0, 0.12), 0 0 2px rgba(0, 0, 0, 0.24)",
      "color": "#000",
      "left": 0,
      "padding": "20px",
      "paddingRight": "48px",
      "position": "fixed",
      "right": 0,
      "top": 0,
      "visibility": "visible",
      "zIndex": 900,
    }
  }
>
  <div
    id="cookie-disclaimer-close"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "fontFamily": "-webkit-pictograph",
        "fontSize": "28px",
        "height": "28px",
        "lineHeight": "28px",
        "marginTop": "-14px",
        "position": "absolute",
        "right": "10px",
        "textAlign": "center",
        "top": "50%",
        "transform": "rotate(45deg)",
        "width": "28px",
      }
    }
  >
    +
  </div>
  <div
    id="cookie-disclaimer-text"
  >
    This website uses cookies to improve service, for analytical and advertising purposes.
    <span
      id="cookie-policy"
      style={
        Object {
          "display": "none",
        }
      }
    >
       By continuing to use the service, you agree to our use of cookies as described in the 
      <a
        href=""
        id="cookie-policy-link"
        style={
          Object {
            "color": "#000",
          }
        }
      >
        Cookie Policy
      </a>
    </span>
  </div>
</div>
`;
